ember-server/src/test/scala/org/http4s/ember/server/EmberServerMtlsSuite.scala

```diff
@@ -0,0 +1,184 @@
+              assert(session.sslSessionId != "")
+              assert(session.cipherSuite != "")
+              assert(session.keySize != 0)
+              assert(session.X509Certificate.isEmpty)
+            }
+
+          Ok("success")
+      }
+      .orNotFound
+  }
+
+  lazy val authTlsClientContext: Resource[IO, TLSContext[IO]] =
+    Resource.eval(Network[IO].tlsContext
+      .fromKeyStoreResource(
+        "keystore.jks",
+        "password".toCharArray,
+        "password".toCharArray
+      )
+    )
+
+  lazy val noAuthClientContext: SSLContext = {
+    val js = KeyStore.getInstance("JKS")
+    js.load(getClass.getResourceAsStream("/keystore.jks"), "password".toCharArray)
+
+    val tmf = TrustManagerFactory.getInstance("SunX509")
+    tmf.init(js)
+
+    val sc = SSLContext.getInstance("TLSv1.2")
+    sc.init(null, tmf.getTrustManagers, null)
+
+    sc
+  }
+
+  lazy val noAuthTlsClientContext: Resource[IO, TLSContext[IO]] =
+    Resource.eval(
+      TLSContext.Builder.forAsync[IO].fromSSLContext(noAuthClientContext).pure[IO]
+    )
+
+  def fixture(tlsParams: TLSParameters, clientTlsContext: Resource[IO, TLSContext[IO]]) =
+    (server(tlsParams), client(clientTlsContext)).mapN(FunFixture.map2(_, _))
+
+  def client(tlsContextResource: Resource[IO, TLSContext[IO]]) = {
+    ResourceFixture(clientResource(tlsContextResource))
+  }
+
+  def clientResource(tlsContextResource: Resource[IO, TLSContext[IO]]) = {
+    (for {
+      tlsContext <- tlsContextResource
+      emberClient <- EmberClientBuilder.default[IO]
+        .withTLSContext(tlsContext)
+        .withoutCheckEndpointAuthentication
+        .build
+    } yield emberClient)
+  }
+
+  def server(tlsParams: TLSParameters) = {
+    ResourceFixture(serverResource(tlsParams))
+  }
+
+  def serverResource(tlsParams: TLSParameters) =
+    (for {
+      tlsContext <- authTlsClientContext
+      emberServer <- EmberServerBuilder
+        .default[IO]
+        .withHttpApp(service[IO])
+        .withTLS(
+          tlsContext,
+          tlsParams
+        )
+        .build
+    } yield emberServer)
+
+  fixture(TLSParameters(needClientAuth = true, protocols = List("TLSv1.2").some), authTlsClientContext).test("Server should handle mTLS request correctly") { case (server, client) =>
+    import org.http4s.dsl.io._
+    import org.http4s.client.dsl.io._
+
+    val uri = Uri
+      .fromString(s"https://${server.address.getHostName}:${server.address.getPort}/dummy")
+      .toOption
+      .get
+    val request = GET(uri)
+    for {
+      response <- client.fetchAs[String](request)
+    } yield assertEquals(clientCertCn, response)
+  }
+
+  fixture(TLSParameters(needClientAuth = true, protocols = List("TLSv1.2").some), noAuthTlsClientContext).test("Server should fail for invalid client auth") { case (server, client) =>
+    client.get(s"https://${server.address.getHostName}:${server.address.getPort}/dummy")(_.status.pure[IO])
+      .intercept[IOException]
```

Comment:
I think you can call `client.status` or similar here?

Reply:
Yes.  If I merge this before it gets cleaned up, it can be a separate PR.  I'm trying to get four releases out the door before I go coach soccer. :smile: